### DISCUSSION TITLE
Question: How to get full transition matrix?

### DISCUSSION THREAD

**Discussion by RoryMMMM** [TheoMichelot/moveHMM#46] [Q&A]
I'm takeing GPS data with some covariates, and running it through a process very similar to the Haggis example. I'm taking the output from the R script and sticking them into a python package for agent based modelling.

The one part I'm sorta stuck on is the transistion matrix when using covariates:

![Image](https://github.com/user-attachments/assets/c4891ef1-7958-4214-8090-856783dfb05d)

![Image](https://github.com/user-attachments/assets/0d116cb9-c104-49d0-86fb-54222565c5fd)

Is there a specific reason for showing the "off-diagonal" matrix instead of the full transistion matrix? Is there a name for this kind of matrix I can look up?

I've taken a quick look at the trMatrix_rcpp example in the ["custom plots"](https://cran.r-project.org/web/packages/moveHMM/vignettes/moveHMM-custom-plots.pdf) pdf but this doesn't seem to give the same kind of values...

**Comment by TheoMichelot** [comment]
Hi, can you please clarify what output you would like? The matrix `mod_3states$mle$beta` has the regression parameters that link the transition probabilities to the covariates. These are the parameters that are estimated, and the transition probabilities can be derived from them for a set of covariate values. You can use the function `predictTPM()` to get transition probabilities based on a fitted model and a data frame of covariate values. In your example, this might look something like:
``` R
# Covariate value(s) at which to evaluate transition probabilities
newData <- data.frame(raster_value = c(0.1, 0.2))
tpm <- predictTPM(m = mod_3states, newData = newData)
```
With this function, you can even get confidence intervals on the transition probabilities using `returnCI = TRUE`.

Let me know if this isn't what you're after.

**Comment by RoryMMMM** [comment]
My workflow looks something like:

- Preprocess some GPS animal tracking data and get some covarate data (land cover, NDVI, canopy cover, etc)
- Fit data to HMM model using MoveHMM
- Implement HMM in Python to do agent based modelling with Repast4Py
- Run Python model and evaluate results 
- Make some changes to data filtering, HMM model, rinse, repeat

I wrote some quick and ugly python code to take the 2 x 6 transition matrix (for a 3 state HMM) and turn it into the full 3x3 transition matrix.  This just makes it easy to calculate the next state given a current state and covariates (in python).

My question is more of a philosophical one because it feels like I'm missing something: Why represent the transition matrix as 2x6 instead of 3x3? Does this make the maths easier somewhere else? Is there a simple way to transform the one to the other?

**Comment by TheoMichelot** [comment]
If you want to compute transition probabilities for new covariate values in Python, you will need to write your own function (or find one in Python) for the transformation. The formula is given in [Michelot et al. (2016)](https://doi.org/10.1111/2041-210X.12578), for example. Briefly, the steps would be:
1. Format the covariates into a design matrix, `covs`, say. The first column is filled with ones (for the intercept), and each subsequent column is for a covariate. 
2. Multiple the design matrix by the matrix of regression coefficients. In R: `covs %*% mod_3states$mle$beta`. This will give you a matrix with one row for each row of the design matrix and one column for each (non-diagonal) transition probability.
3. For each row of the matrix from step 2, you then need to compute the corresponding transition probability matrix, using the formula from the paper. This requires adding zeros for the missing diagonal elements, splitting by row of the transition probability matrix, and then applying the multinomial logit link / softmax function to the elements of each row of the transition probability matrix.

All of this happens inside `predictTPM()` or `trMatrix_rcpp()` in the package. Let me know if anything is unclear.
